src/simplemmap.c

```diff
@@ -365,8 +365,146 @@ static void InitMmapRealClass(DllInfo *dll)
  * Constructor
  */
 
-#ifdef Win32
-# error "I'm sure this needs adjusting for Windows, so punt for now."
+//#define DEBUG_PRINT(x) REprintf(x);
+#define DEBUG_PRINT(x) do { } while (0)
+
+#define MMAP_FILE_WARNING_OR_ERROR(str, ...) do {   \
+    if (warn) {                 \
+        warning(str, __VA_ARGS__);          \
+        return NULL;                \
+    }                       \
+    else error(str, __VA_ARGS__);           \
+    } while (0)
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+static void mmap_finalize(SEXP eptr)
+{
+    DEBUG_PRINT("finalizing ... ");
+    void *p = R_ExternalPtrAddr(eptr);
+    R_SetExternalPtrAddr(eptr, NULL);
+    if (p != NULL) {
+        UnmapViewOfFile(p); /* don't check for errors */
+        R_SetExternalPtrAddr(eptr, NULL);
+    }
+    DEBUG_PRINT("done\n");
+}
+
+static LPSTR GetErrorMessage()
+{
+    static CHAR buf[256];
+    FormatMessageA(
```

Comment:
Using `FormatMessageW` would be nice here, but given that the default locale is Windows-1252 there might be no point?